###Describing SVM for Regression

Given a constant $ \epsilon $ and a set of training points, SVM linear regression tries to find a linear function **'f'**  in such a way that each of the training label **y** (for each of the training point **x**) deviates from **f(x)** by atmost $ \epsilon $  and at the same time is as flat as possible. The flatness here refers to the complexity of the model. A less complex model will not overfit the data. If the model is flat it will be good in generalizing.



In other words, it tries to find the hyperplane that holds maximum training observations within the margin $\epsilon $ (tolerance level), in such a way that it is as flat as possible. But it is possible that not all the points fall within the tolerance level.




For points that lie outside the $\epsilon $ margin (i.e., where the prediction error is greater than $\epsilon $), slack variables allow the model to tolerate some degree of error. Those slack variables are denoted as $ \xi $
 and $ \xi $*. These deviations from  margin are penalized by appropriate regularization parameter C.






 So SVR finds the line that maximizes the margin (makes it flat) while minimizing slack violations and keeping as many points as possible within the epsilon-tube.






A higher C value will try to minimize the slack variables (i.e., it will try to fit as many points as possible inside the ε margin, even if that leads to a more complex model).
A lower C value means the model will tolerate more slack (i.e., it won’t force all points to be within the margin, potentially resulting in a flatter, simpler model).


By default epsilon value is taken as 0.1 and C is taken to be 1.0.

Hyperparameter tuning of  $ \epsilon $ and C can be performed using GridSearch   





In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
#Reading preprocessed training and test sets from our linear regression assignment.
import pandas as pd

housing_prepared_train = pd.read_csv("/content/drive/MyDrive/Machine_Learning/Linear_Regression/housing_prepared_train.csv")
housing_labels_train = pd.read_csv("/content/drive/MyDrive/Machine_Learning/Linear_Regression/housing_labels_train.csv")
housing_prepared_test = pd.read_csv("/content/drive/MyDrive/Machine_Learning/Linear_Regression/housing_prepared_test.csv")
housing_labels_test = pd.read_csv("/content/drive/MyDrive/Machine_Learning/Linear_Regression/housing_labels_test.csv")

print(housing_prepared_train.shape)
print(housing_prepared_test.shape)
print(housing_labels_train.shape)
print(housing_labels_test.shape)


(16512, 16)
(4128, 16)
(16512, 1)
(4128, 1)


In [ ]:
parameters = {
    "C":[0.01,0.1,1.0,10.0,1000.0,10000.0],
    "epsilon":[0.01,0.1,1.0,10.0,100.0],
}

In [ ]:
from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVR

grid_search = GridSearchCV(estimator=SVR(kernel="linear"), param_grid=parameters, cv=3,n_jobs=-1)


In [ ]:
grid_search.fit(housing_prepared_train,housing_labels_train)

/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


GridSearchCV(cv=3, estimator=SVR(kernel='linear'), n_jobs=-1,
             param_grid={'C': [0.01, 0.1, 1.0, 10.0, 1000.0, 10000.0],
                         'epsilon': [0.01, 0.1, 1.0, 10.0, 100.0]})

In [ ]:
print(grid_search.best_params_)

{'C': 10000.0, 'epsilon': 10.0}


In [ ]:

svm_reg = grid_search.best_estimator_


In [ ]:
# Get the predictions from SVR
predictions = svm_reg.predict(housing_prepared_test)

In [ ]:
# Let us print the first 5 predictions

predictions_df = pd.DataFrame(predictions)
predictions_df.head()

0
0  153014.050606
1   92860.694461
2  187643.281574
3  259678.344422
4  102980.703397

In [ ]:
# Let us print the root mean square error of our predictions
from sklearn.metrics import mean_squared_error
import numpy as np

mse = mean_squared_error(housing_labels_test, predictions)
rmse = np.sqrt(mse)
print(rmse)

69373.26081118616


The root mean square error is found to be 69373.26081118616